swaybar/render.c

```diff
@@ -681,6 +732,76 @@ static uint32_t render_workspace_button(struct render_context *ctx,
 	return output->height;
 }
 
+uint32_t render_focused_window_icon(cairo_t *cairo,
+		struct swaybar_output *output,
+		double *x) {
+	assert(output);
+	assert(output->bar);
+
+	if (!output->bar->focused_window) {
+		return output->height;
+	}
+
+	uint32_t height = output->height * output->scale;
+	int padding = 4;
+	int target_size = height - 2 * padding;
+
+	char *icon_name = output->bar->focused_window->icon_name;
+	cairo_surface_t *icon = NULL;
+	if (icon_name) {
+		char *icon_theme = output->bar->config->icon_theme;
+		list_t *basedirs = get_basedirs();
+		int min_size = 0;
+		int max_size = 0;
+
+		list_t *themes = create_list();
+		// TODO: Load correct theme
+		list_add(themes, "Adwaita");
+
+		assert(output->bar->tray);
+		assert(output->bar->tray->themes);
+		char *icon_path = find_icon(output->bar->tray->themes,
+				basedirs,
+				icon_name,
+				target_size,
+				icon_theme,
+				&min_size,
+				&max_size);
+		icon = load_background_image(icon_path);
+	} else {
+		// TODO: Generate a image on the fly
+		icon = load_background_image(
+				"/usr/share/icons/Adwaita/16x16/apps/"
+				"utilities-terminal-symbolic.symbolic.png");
```

Comment:
If the application doesn't have the icon set, I think it's better to not show it at all instead of hard-coding this icon which also introduces problems for people who don't have Adwaita theme.

Reply:
Yes, my plan was to generate a image with cairo. Like it is done with missing tray icons.

Reply:
In the case of tray, it makes sense to have a dummy icon because otherwise it won't be useful. But here it would be better if application icon is not shown at all if the application doesn't define any.